### PR TITLE
Add DataConfig and split-based dataloaders

### PR DESCRIPTION
**Pull request by aaishwarymishra**
This pull request introduces a significant refactor to how datasets and dataloaders are configured and instantiated, moving from a monolithic dataset configuration to a more flexible, split-based approach. The changes enable separate configuration of training and validation splits, including their datasets and dataloaders, and update the codebase to support this new structure. Additionally, the dataset builder and registry are updated for improved modularity and extensibility.

**Refactor to split-based data configuration:**

* Replaces the old single `dataset` config with a new `data` section in `ProjectConfig`, allowing separate configuration of `train` and `val` splits, each with their own dataset and dataloader settings. Updates related config classes (`DataConfig`, `SplitConfig`, `DataLoaderConfig`) to support this structure. (`trainite/config/base.py`, `trainite/config/__init__.py`, `trainite/cli/init.py`, `config.yaml`, `trainite/templates/project/README.md`) [[1]](diffhunk://#diff-73352dfb97a0b9afb3ad346b22af2fdb5df6e72c73da6aa0df96d85c0a4c0cddL25-R50) [[2]](diffhunk://#diff-c90462a31c78a843f174e06c2c564385aa3cd0a5d38fdd758b926f63f0ce5d1cR3-R7) [[3]](diffhunk://#diff-c90462a31c78a843f174e06c2c564385aa3cd0a5d38fdd758b926f63f0ce5d1cR33-R35) [[4]](diffhunk://#diff-1e20c19b0f95a48591a0f35c025ef834b85ff6c4deaecb675a711f8499f1bf40L11-R19) [[5]](diffhunk://#diff-1e20c19b0f95a48591a0f35c025ef834b85ff6c4deaecb675a711f8499f1bf40L293-R338) [[6]](diffhunk://#diff-d8d0422389f03d783e32e627250fe29834bd09c6361640d1ff00661dd6820034L1-L38) [[7]](diffhunk://#diff-ada0f955d81307b406b77a5ab5e7423974538138ae1ead8e4005cce3b61cd101L42-R46)
* Updates the trainer (`PreTrainer`) to build dataloaders from the new split-based config, with logic to fall back to the training split if a validation split is not provided, and logs a warning in that case. (`trainite/trainers/pretrainer.py`) [[1]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caL17-R34) [[2]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caL49-R73)

**Dataset builder and registry improvements:**

* Refactors the string reverse dataset builder to return a single dataset object instead of a tuple of dataloaders, and updates its registration in the dataset registry accordingly. The dataset config now includes a single `size` parameter instead of separate `train_size` and `val_size`. (`trainite/datasets/string_reverse.py`, `trainite/config/dataset.py`, `trainite/config/registry.py`) [[1]](diffhunk://#diff-e0dac8bceead0c2b475a6440a83d3cb3f966218558961b9832d2eeb75dbb4250L84-L126) [[2]](diffhunk://#diff-cc556a921ff67b5954e6a97505d0e02344c19dcd16e3cd057473afd7355b07eeL2-L18) [[3]](diffhunk://#diff-f1e4797054e117a0de73729f410d0b1781a627ce2ff13b9d02d2cbf0e5fc1a4bL1-R29) [[4]](diffhunk://#diff-f1e4797054e117a0de73729f410d0b1781a627ce2ff13b9d02d2cbf0e5fc1a4bL42-L43) [[5]](diffhunk://#diff-f1e4797054e117a0de73729f410d0b1781a627ce2ff13b9d02d2cbf0e5fc1a4bL53-R49) [[6]](diffhunk://#diff-f1e4797054e117a0de73729f410d0b1781a627ce2ff13b9d02d2cbf0e5fc1a4bL65-L73)

**Utility and API enhancements:**

* Adds a generic `get_target` utility for resolving a class or function from a string path, used to instantiate collate functions and other components dynamically. (`trainite/utils.py`)

**Minor cleanups and consistency:**

* Removes unused or redundant parameters from dataset and trainer configs, and updates naming for clarity and consistency. (`trainite/config/trainer.py`, `trainite/config/dataset.py`) [[1]](diffhunk://#diff-b17caca894810164f85d88750aff84f166bdcafaeebe3c796ca521980e1d18bdL5) [[2]](diffhunk://#diff-cc556a921ff67b5954e6a97505d0e02344c19dcd16e3cd057473afd7355b07eeL2-L18)

These changes collectively make the configuration system more modular, extensible, and user-friendly, especially for projects needing custom splits or dataloader settings.Replace legacy dataset config with DataConfig for train and val Dataset builders return Dataset instances instead of dataloader pairs PreTrainer builds DataLoader from SplitConfig and supports per-split collate_fn. Registry exposes builder and collate symbols and utils.get_target
was added. Remove example config.yaml and update README
[Copilot is generating a summary...]

### CODE REVIEW

File: trainite/utils.py [hunk 1]
```diff
@@ -35,3 +35,22 @@ def instantiate(config: ComponentConfig, **kwargs) -> Any:
     final_kwargs = {**params, **kwargs}
 
     return target_symbol(**final_kwargs)
+
+
+def get_target(target_path: str) -> Any:
```
Comment (TahaZahid05) [inline trainite/utils.py]: Can we refactor this as it is quite similar to `instantiate()` function definition.

File: trainite/config/registry.py [hunk 2]
```diff
@@ -1,35 +1,32 @@
...
+from pydantic import BaseModel
+
 from trainite.config.dataset import StringReverseDatasetConfig
 from trainite.config.model import TransformerModelConfig
 from trainite.config.trainer import PreTrainerConfig
 
 
-@dataclass(frozen=True)
-class ComponentSpec:
+class ComponentSpec(BaseModel):
```
Comment (TahaZahid05) [inline trainite/config/registry.py:10]: I do not understand this change. can you explain please?

File: config.yaml
Comment (TahaZahid05) [inline config.yaml:1]: Why remove the file?

File: trainite/trainers/pretrainer.py [hunk 3]
```diff
@@ -14,13 +14,24 @@
...
+from torch.utils.data import DataLoader
+
+from trainite.config import ProjectConfig, SplitConfig, dump_config
+from trainite.utils import instantiate, get_target
 
 logger = logging.getLogger(__name__)
 
 
 class PreTrainer:
+    def _build_dataloader(self, split_config: SplitConfig) -> DataLoader:
```
Comment (TahaZahid05) [inline trainite/trainers/pretrainer.py]: let's move this below __init__

File: trainite/cli/init.py [hunk 4]
```diff
@@ -290,17 +298,44 @@ def init_project(args: argparse.Namespace) -> None:
...
-    dataset_component.target = (
-        f"dataset.{dataset_spec.name}.{dataset_spec.builder_symbol}"
-    )
+
+    # Build train split
+    train_dataset = dataset_spec.config_cls()
+    train_dataset.target = f"dataset.{dataset_spec.name}.{dataset_spec.builder_symbol}"
+
+    # Build val split
+    val_dataset = dataset_spec.config_cls()
```
Comment (TahaZahid05) [inline trainite/cli/init.py:310]: We should change the seed, and the dataset's size as currently it is identical to train_dataset?

File: config.yaml
Comment (aaishwarymishra) [reply config.yaml:1]: It was never being used

File: trainite/cli/init.py (hunk 4 above)
Comment (aaishwarymishra) [reply trainite/cli/init.py:310]: dataset size and seed is subjective, not all dataset config will have it , some can simply load from the internet, user should edit the validation dataset according to their choice, we have mentioned it on the readme.

File: trainite/config/registry.py (hunk 2 above)
Comment (aaishwarymishra) [reply trainite/config/registry.py:10]: changed to pydantic classes for consitency , as we are using pydantic for validation so why not use it here too to remain consistent.

File: config.yaml
Comment (TahaZahid05) [reply config.yaml:1]: > It was never being used

It is helpful for us if we want to quickly test our functionality using `main.py` and `config.yaml` rather than make an experiment. I think we should keep it for development.

Comment (aaishwarymishra) [reply config.yaml:1]: I wont recommend it as it wont work for other dataset, trainer model. Also we need  to manually update everytime there is code change. init should be our main way to test.

File: trainite/config/registry.py (hunk 2 above)
Comment (TahaZahid05) [reply trainite/config/registry.py:10]: Makes sense. Though, we should still freeze the class like the dataclass did so we don't accidentally change it in code. We can use `model_config = ConfigDict(frozen=True)`

Comment (aaishwarymishra) [reply trainite/config/registry.py:10]: done.

**Review by TahaZahid05 — APPROVED** [review]
